### PR TITLE
Add Duration Calculator

### PR DESCRIPTION
**Pull request by sam09**
###### New Duration Calculator Goodie

Choose the most relevant items and use the following title template to name
your Pull Request.

- [x] New Instant Answer
    - [x] Other: **`New Duration Calculator Instant Answer`**
- [x] Non–Instant Answer
    - [x] Other (Tests}: 

###### Description of changes

Add a new IA to answer queries asking for addition or subtraction of two time quantities.

@GuiltyDolphin 

---

Instant Answer Page: https://duck.co/ia/view/duration_calculator

### CODE REVIEW

**Comment by daxtheduck** [comment]
@sam09 Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by zekiel** [comment]
Thanks for this @sam09 We'll take a look and leave you some feedback -- looks like it's failing the travis build.

**Comment by gaulrobe** [comment]
@sam09 it may be worthwhile looking into leveraging http://search.cpan.org/~drolsky/DateTime-1.28/lib/DateTime/Duration.pm for these calculations. Thoughts?

**Comment by sam09** [comment]
@gaulrobe I will look at it

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+my $operation_re = qr/(?<operand1>$operand_re)$action_re\s+(?<operand2>$operand_re)?/i;
+
+my $what_re = qr/what ((is|was|will) the )?/i;
+my $time_re = qr/(?<time>time)/i;
+my $will_re = qr/ (was it|will it be|is it|be)/i;
+my $full_time_regex = qr/^($what_re?$time_re$will_re? )?($operation_re)[\?.]?$/i;
+
+
+
+sub format_result {
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: I think using `DateTime::Format::Human::Duration` would be good here - that'll automatically sort out things like `3 hours, 7 minutes, and 10 seconds` from a Duration object.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 2]
```diff
@@ -0,0 +1,185 @@
+package DDG::Goodie::DurationCalculator;
+# ABSTRACT: Finds duration between two times
+
+use strict;
+use DDG::Goodie;
+with 'DDG::GoodieRole::Dates';
+with 'DDG::GoodieRole::NumberStyler';
+use DateTime::Duration;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: You've already got it imported - may as well use it for the math ;)

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+triggers any => qw(seconds minutes hours);
+triggers any => qw(plus minus + - before after);
+
+my $number_re = number_style_regex();
+my $action_re = qr/(?<action>plus|add|\+|\-|minus|subtract)/i;
+
+my $day_re = qr/(?<day>$number_re)\s+(day|days)/i;
+my $hour_re = qr/(?<hour>$number_re)\s+(hour|hours)/i;
+my $min_re = qr/(?<min>$number_re)\s+(minute|minutes)/i;
+my $sec_re = qr/(?<sec>$number_re)\s+(second|seconds)/i;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: Lets have a single regex for the units - then you can parse it similar to https://github.com/duckduckgo/zeroclickinfo-goodies/blob/4cde720b628c819ab1f03f90bcd8b3bb45423a3e/lib/DDG/GoodieRole/Dates/Parser.pm#L488 and just pass all the units to `DateTime::Duration`.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+
+zci answer_type => 'duration_calculator';
+
+# Caching - http://docs.duckduckhack.com/backend-reference/api-reference.html#caching`
+zci is_cached => 1;
+
+# Triggers - http://docs.duckduckhack.com/walkthroughs/calculation.html#triggers
+triggers any => qw(second minute hour);
+triggers any => qw(seconds minutes hours);
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: What about days, months, and years?

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	}
+    return $string;
+}
+sub get_action_for {
+    my $action = shift;
+    return '+' if $action =~ /^(\+|plus|from|in|add|after)$/i;
+    return '-' if $action =~ /^(\-|minus|ago|subtract|before)$/i;
+}
+
+sub get_values{
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: As mentioned above, the `_util...` helpers (or similar) may be useful here.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+    $day = "0" unless defined $day;
+    $hour = "0" unless defined $hour;
+    $min = "0" unless defined $min;
+    $sec = "0" unless defined $sec;
+    
+    return ($day, $hour, $min, $sec);
+}
+
+sub add{
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: As @gaulrobe mentioned - we shouldn't be doing any of the calculating ourselves, there's plenty of packages that can do that for us.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+my $min_re = qr/(?<min>$number_re)\s+(minute|minutes)/i;
+my $sec_re = qr/(?<sec>$number_re)\s+(second|seconds)/i;
+
+my $operand_re = qr/($day_re\s*)?($hour_re\s*)?($min_re\s*)?($sec_re\s*)?/i;
+my $operation_re = qr/(?<operand1>$operand_re)$action_re\s+(?<operand2>$operand_re)?/i;
+
+my $what_re = qr/what ((is|was|will) the )?/i;
+my $time_re = qr/(?<time>time)/i;
+my $will_re = qr/ (was it|will it be|is it|be)/i;
+my $full_time_regex = qr/^($what_re?$time_re$will_re? )?($operation_re)[\?.]?$/i;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: We'll want to get rid of these - they are very specific to DateMath (`what will the time be...`).

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+	}
+	else {
+		$string .= " second";
+	}
+    return $string;
+}
+sub get_action_for {
+    my $action = shift;
+    return '+' if $action =~ /^(\+|plus|from|in|add|after)$/i;
+    return '-' if $action =~ /^(\-|minus|ago|subtract|before)$/i;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: We can lose `from`, `in`, `after`, `ago`, and `before`.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    	$day = $day - 1;
+    }
+    return format_result($day, $hour, $min, $sec);
+}
+
+sub get_result {
+    my ($values1, $values2, $action) = @_;
+    
+    return  add (\@$values1, \@$values2) if $action eq '+';
+    return subtract (\@$values1, \@$values2) if $action eq '-';
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: Why are you reffing a deref?

**Comment by sam09** [comment]
@GuiltyDolphin @gaulrobe  It seems that DateTime::Duration has issues with subtraction. A query like 1 minute - 25 seconds gives 1 minute 25 seconds as answer

File: lib/DDG/Goodie/DurationCalculator.pm (hunk 3 above)
Comment (sam09) [inline lib/DDG/Goodie/DurationCalculator.pm]: Hey i think adding a single regex for unit would mean that we cannot differentiate between different time units. For example, a query like 7 days 2 days is not valid. However having same regex would make it impossible to differentiate such terms.

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: Then just return if a unit appears twice; I'm assuming you mean cases like `1 hour + 7 days, 2 days, 3 hours`?

Just check whatever you are collecting the units into (probably a hash) that there is no current value for that particular unit.

**Comment by GuiltyDolphin** [comment]
@sam09 Can you throw me an example? It's working fine for me (you may have to use `format_duration_between`):

```perl

use DateTime;
use DateTime::Duration::Format::Human;

my $span = DateTime::Format::Human::Duration->new();

my $start = DateTime->now();
my $end  = $start->clone->add(
    minutes => 1,
    seconds => -25,
);

$span->format_duration_between($start, $end);
# 35 seconds
```

**Comment by sam09** [comment]
@gaulrobe @GuiltyDolphin I fixed most of the issues

File: t/DurationCalculator.t [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+sub build_structured_answer {
+    my @test_params = @_;
+
+    return $test_params[0],
+        structured_answer => {
+
+            data => {
+                title    => $test_params[0],
+                # subtitle => "My Subtitle",
+                # image => "http://website.com/image.png",
```
Comment (gaulrobe) [inline t/DurationCalculator.t]: You can get rid of these two lines if there will be no `subtitle` or `image`.

File: t/DurationCalculator.t [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            },
+
+            templates => {
+                group => "text",
+                # options => {
```
Comment (gaulrobe) [inline t/DurationCalculator.t]: Same here, can remove this commented out code 👍

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+zci answer_type => 'duration_calculator';
+
+# Caching - http://docs.duckduckhack.com/backend-reference/api-reference.html#caching`
+zci is_cached => 1;
+
+# Triggers - http://docs.duckduckhack.com/walkthroughs/calculation.html#triggers
+triggers any => qw(second minute hour day);
+triggers any => qw(seconds minutes hours days);
+triggers any => qw(plus minus + - before after);
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: We don't need 'before' or 'after'.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    
+    my $start = DateTime->now();
+    my $end  = $start->clone->add($result);
+
+    return $span->format_duration_between($start, $end);
+}
+sub get_action_for {
+    my $action = shift;
+    return '+' if $action =~ /^(\+|plus)$/i;
+    return '-' if $action =~ /^(\-|minus)$/i;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: Need to add 'add' and 'subtract'.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+sub get_values{
+    my ($operand) = @_;
+  
+    my %modifiers;
+    while ($operand =~ /$term_re/g) {
+        my $number = $+{number};
+        my $unit = $+{unit};
+        
+        if( substr($unit, -1) ne 's') {
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: `$unit .= 's' unless $unit =~ /s$/;`

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    
+    
+    $action = get_action_for($action);
+    my $result = get_result ($dur1, $dur2, $action);
+
+    return $result,
+            structured_answer => {
+
+            data => {
+                title    => $result,
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: We should have a subtitle with the operations.

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: And the title should have the actual result :)

File: t/DurationCalculator.t [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+my $add_query = "2 day 7 minutes 39 seconds + 1 day 9 hours 58 minutes 25 seconds";
+my $sub_query = "1 day 9 hours 59 minutes 14 seconds - 1 day 9 hours 58 minutes 25 seconds";
+my $invalid_query = "2 day 2 day + 4 day";
+ddg_goodie_test(
+    [qw( DDG::Goodie::DurationCalculator )],
+    
+    
+    $add_query => build_test("3 days, 10 hours, 6 minutes, and 4 seconds"),
+    $sub_query => build_test("49 seconds"),
```
Comment (GuiltyDolphin) [inline t/DurationCalculator.t]: We should be testing both the formatting of the query *and* the produced result.

File: lib/DDG/Goodie/DurationCalculator.pm (hunk 15 above)
Comment (sam09) [inline lib/DDG/Goodie/DurationCalculator.pm]: So the subtitle should have a sort of query. Like what Datemath has?

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: So something like:

If the query were "1 hour + 3 hours plus 2 hours"

Then you might have:

Subtitle: "1 hour + 3 hours + 2 hours"
Title: "6 hours"

File: t/DurationCalculator.t [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    $invalid_query => undef,
```
Comment (GuiltyDolphin) [inline t/DurationCalculator.t]: Lets add a test for something like `3 days + 1 hour in hours` (`= 73 hours`).

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+
+
+zci answer_type => 'duration_calculator';
+
+# Caching - http://docs.duckduckhack.com/backend-reference/api-reference.html#caching`
+zci is_cached => 1;
+
+# Triggers - http://docs.duckduckhack.com/walkthroughs/calculation.html#triggers
+triggers any => qw(second minute hour day);
+triggers any => qw(seconds minutes hours days);
```
Comment (moollaza) [inline lib/DDG/Goodie/DurationCalculator.pm]: I think we should also add `hrs`, `mins`, `secs`?

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: And `weeks`, `months`, `years`.

I imagine that this sort of thing may well be used for high-precision units like milliseconds (maybe even micro or nano); we might want to look into which of those we *could* support.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+triggers any => qw(plus minus + - before after);
+
+my $number_re = number_style_regex();
+my $action_re = qr/(?<action>plus|add|\+|\-|minus|subtract)/i;
+my $unit_re = qr/(day|hour|minute|second)s?/i;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: Considering we're going to be repeating units all over the place, I think we should pull out the supported units into a var that we can reference - we should be able to manipulate them fairly easily then.

File: t/DurationCalculator.t [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+sub build_test { test_zci(build_structured_answer(@_)) }
+
+my $add_query = "2 day 7 minutes 39 seconds + 1 day 9 hours 58 minutes 25 seconds";
+my $sub_query = "1 day 9 hours 59 minutes 14 seconds - 1 day 9 hours 58 minutes 25 seconds";
+my $invalid_query = "2 day 2 day + 4 day";
+ddg_goodie_test(
+    [qw( DDG::Goodie::DurationCalculator )],
+    
+    
+    $add_query => build_test("3 days, 10 hours, 6 minutes, and 4 seconds"),
```
Comment (GuiltyDolphin) [inline t/DurationCalculator.t]: Also we're going to need more tests :)

File: lib/DDG/Goodie/DurationCalculator.pm (hunk 18 above)
Comment (sam09) [inline lib/DDG/Goodie/DurationCalculator.pm]: ``DateTime::Duration``  supports ``years`` ``months`` ``weeks`` and ``nanoseconds``

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DurationCalculator.pm]: Hmm, well it's your call on `nanoseconds` (seeing as we don't have `milli` and `micro`); but I think `years`, `months`, and `weeks` would be great :+1:

**Comment by sam09** [comment]
@zekiel  @gaulrobe  @GuiltyDolphin @moollaza Thanks a lot for your feedback.. Honestly speaking, I have enjoyed contributing to DuckDuckGo.

File: lib/DDG/Goodie/DurationCalculator.pm [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    
+    $action = get_action_for($action);
+    my $result = get_result ($dur1, $dur2, $action);
+    my $format_query = $operand1.$action." ".$operand2;
+    return $result,
+            structured_answer => {
+
+            data => {
+                title    => $result,
+                subtitle => $format_query,
```
Comment (moollaza) [inline lib/DDG/Goodie/DurationCalculator.pm]: This should also indicate the operation: "Duration Math: 30 days + 2 weeks"

**Comment by moollaza** [comment]
@sam09 looks like we need to add some handling for negative results?  Or are we purposely not showing it as a negative number?

https://beta.duckduckgo.com/?q=1+day+-+45+hours&ia=answer

**Comment by moollaza** [comment]
@sam09 Thanks! This LGTM :+1:

**Comment by sam09** [comment]
@moollaza negative results are avoided purposefully.

**Comment by moollaza** [comment]
@sam09 I think we need to indicate negative hours. If the user doesn't know if the result should be positive or negative it becomes hard to understand.

https://beta.duckduckgo.com/?q=60+days+-+900+hours&ia=answer

e.g. Does that mean 900 hours is 3 weeks less than 60 days? or 3 weeks longer than 60 days?

**Comment by sam09** [comment]
@moollaza Yeah I agree to that.. Now that this branch is merged, I am guessing we should probably open another issue for this. Your thoughts on that?

**Comment by GuiltyDolphin** [comment]
@sam09 Feel free to open another PR with additional changes :) we can discuss anything else that might need doing there.

**Pull request merged**